pkg/agent/storage/gcs.go

```diff
@@ -0,0 +1,132 @@
+package storage
+
+import (
+	gstorage "cloud.google.com/go/storage"
+	"context"
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/googleapis/google-cloud-go-testing/storage/stiface"
+	"google.golang.org/api/iterator"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+type GCSProvider struct {
+	Client stiface.Client
+}
+
+func (p *GCSProvider) DownloadModel(modelDir string, modelName string, storageUri string) error {
+	log.Info("Downloading model ", "modelName", modelName, "storageUri", storageUri, "modelDir", modelDir)
+	gcsUri := strings.TrimPrefix(storageUri, string(GCS))
+	tokens := strings.SplitN(gcsUri, "/", 2)
+	prefix := ""
+	if len(tokens) == 2 {
+		prefix = tokens[1]
+	}
+	ctx := context.Background()
+	gcsObjectDownloader := &GCSObjectDownloader {
+		Context: ctx,
+		StorageUri: storageUri,
+		ModelDir: modelDir,
+		ModelName: modelName,
+		Bucket: tokens[0],
+		Item: prefix,
+	}
+	it, err := gcsObjectDownloader.GetObjectIterator(p.Client)
+	if err != nil {
+		return fmt.Errorf("unable to get object iterator because: %v", err)
+	}
+	if err := gcsObjectDownloader.Download(p.Client, it); err != nil {
+		return fmt.Errorf("unable to download object/s because: %v", err)
+	}
+	return nil
+}
+
+type GCSObjectDownloader struct {
+	Context    context.Context
+	StorageUri string
+	ModelDir   string
+	ModelName  string
+	Bucket     string
+	Item       string
+}
+
+func (g *GCSObjectDownloader) GetObjectIterator(client stiface.Client) (stiface.ObjectIterator, error) {
+	query := &gstorage.Query{Prefix: g.Item}
+	return client.Bucket(g.Bucket).Objects(g.Context, query), nil
+}
+
+func (g *GCSObjectDownloader) Download(client stiface.Client, it stiface.ObjectIterator) error {
+	var errs []error
+	// flag to help determine if query prefix returned an empty iterator
+	var foundObject = false
```

Comment:
I added in a flag to determine if the returned iterator is empty (meaning that the storage uri prefix used to search for objects returned nothing). If the flag is still false after iteration, that means no objects were found and we should log a warning for the user. This was done since the object iterator does not have a way to check size. That said I'm not entirely happy with this solution - any suggestions?